### ISSUE TITLE
Performance drops when changing the controller

### ISSUE THREAD

**Issue by jli111** [horovod/horovod#3868]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet) Tensorflow 
2. Framework version: 2.7.1
3. Horovod version: 0.26
4. MPI version: Open MPI v5.0 main branch
5. CUDA version: 11.2
6. NCCL version: 2.8
7. Python version: 3.8
8. Spark / PySpark version:
9. Ray version:
10. OS and version:
11. GCC version: 9.3
12. CMake version: 3.23

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Bug report:**
Please describe erroneous behavior you're observing and steps to reproduce it.

I am encountering unexpected performance drop when comparing MPI as controller with Gloo as controller. Would like to seek any advice/tip to understand this behavior. Thanks in advance!

So I have built two Horovod with differnt controllers in different conda vitual environments, for both cases NCCL is configured to handle all GPU collective operations, the only difference is just relying on either MPI or Gloo to handle CPU operations and otehr stuff. The command line of building Horovod against Gloo (env_gloo) is:

`HOROVOD_WITH_GLOO=1  HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITHOUT_MPI=1 HOROVOD_WITH_NCCL=1 HOROVOD_WITHOUT_PYTORCH=1 HOROVOD_WITHOUT_MXNET=1 HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_GPU_ALLGATHER=NCCL HOROVOD_GPU_BROADCAST=NCCL HOROVOD_GPU_ALLTOALL=NCCL HOROVOD_GPU_REDUCESCATTER=NCCL python setup.py sdist
`
then
`HOROVOD_WITH_GLOO=1  HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITHOUT_MPI=1 HOROVOD_WITH_NCCL=1 HOROVOD_WITHOUT_PYTORCH=1 HOROVOD_WITHOUT_MXNET=1 HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_GPU_ALLGATHER=NCCL HOROVOD_GPU_BROADCAST=NCCL HOROVOD_GPU_ALLTOALL=NCCL HOROVOD_GPU_REDUCESCATTER=NCCL pip install --no-cache-dir ./dist/horovod-0.26.1.tar.gz horovod`

Results in:
```
Horovod v0.26.1:

Available Frameworks:
    [X] TensorFlow
    [X] PyTorch
    [ ] MXNet

Available Controllers:
    [ ] MPI
    [X] Gloo

Available Tensor Operations:
    [X] NCCL
    [ ] DDL
    [ ] CCL
    [ ] MPI
    [X] Gloo
```

While the command line of build Horovod against MPI (env_mpi) is:
` HOROVOD_WITHOUT_GLOO=1  HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITH_MPI=1 HOROVOD_WITH_NCCL=1 HOROVOD_WITHOUT_PYTORCH=1 HOROVOD_WITHOUT_MXNET=1 HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_GPU_ALLGATHER=NCCL HOROVOD_GPU_BROADCAST=NCCL HOROVOD_GPU_ALLTOALL=NCCL HOROVOD_GPU_REDUCESCATTER=NCCL python setup.py sdist`
and
`HOROVOD_WITHOUT_GLOO=1  HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITH_MPI=1 HOROVOD_WITH_NCCL=1 HOROVOD_WITHOUT_PYTORCH=1 HOROVOD_WITHOUT_MXNET=1 HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_GPU_ALLGATHER=NCCL HOROVOD_GPU_BROADCAST=NCCL HOROVOD_GPU_ALLTOALL=NCCL HOROVOD_GPU_REDUCESCATTER=NCCL pip install --no-cache-dir ./dist/h
orovod-0.26.1.tar.gz horovod`

check build:
```
Horovod v0.26.1: 
Available Frameworks:
    [X] TensorFlow
    [X] PyTorch
    [ ] MXNet 

Available Controllers:
    [X] MPI
    [ ] Gloo

Available Tensor Operations:
    [X] NCCL
    [ ] DDL
    [ ] CCL
    [X] MPI
    [ ] Gloo
```

The comparison experiments were executed across 4 GPU nodes. Each node contains 2 IBM POWER9 CPUs and 6 NVIDIA V100 GPUs. The internode connection goes through Dual-port Mellanox EDR InfiniBand network. The training is scaled by 24 processes mapped separately on 24 GPUs.

The comparison experiments are submitted in the same job script with same environment variables, and the marjor part of the LSF job script is like:
```
NUM_NODE=4
PROC_PER_NODE=6
NUM_PROC=$(($NUM_NODE*$PROC_PER_NODE))

export HOROVOD_GLOO_IFACE="ib0"
export NCCL_SOCKET_IFNAME="ib0"
export GLOO_SOCKET_IFNAME="ib0"

conda activate env_mpi

mpirun -np $NUM_PROC -H $hosts  --map-by ppr:6:node -x PATH -x LD_LIBRARY_PATH python test_mnist.py

conda deactivate
conda activate env_gloo

horovodrun -np $NUM_PROC -H $hosts  --network-interface ib0 python test_mnist.py
```
Since I notice huge difference in total execution time, I have tracked the total count of NCCL Allreduce invokation and the aggreageted cost of NCCL Allreduce at[ here ](https://github.com/horovod/horovod/blob/master/horovod/common/ops/nccl_operations.cc#L185)

MNIST test (batch size 128, 10 epochs per run):
|      | Total execution cost (s) |  NCCL Allreduce cost (ms) | NCCL Allreduce cnout |
| ------------- |:-------------:| :-------------: | :--------: |
| MPI (1st run)  | 15.02 | 1355 | 1119 |
| Gloo (1st run) |  24.18 |  3039 | 872 |
| MPI (2nd run)  |  15.06| 1352 | 1116 |
| Gloo (2nd run) |  13.13  | 1346 | 804 |
| MPI (3rd run)  | 15.27 | 1382 | 1120 |
| Gloo (3rd run) | 13.25 |  1302 | 858 |

If we take a closer look at each epoch cost of one run:

MPI: 4.514014959, 0.8593580723 ,0.8258671761, 0.8176670074, 0.8071436882, 0.8929388523, 0.8544001579, 0.8469939232, 0.8205637932, 0.8525788784;
Gloo: 3.911006212, 0.8106968403, 0.7058227062, 0.7383961678, 0.7432246208, 0.729790926, 0.7165548801, 0.710360527, 0.7231853008, 0.7140107155;

The sum of all epochs for MPI is 12.09 s, while the sum for Gloo is 10.5 s. There are 3ish overhead for both cases, since I have commented out the checkpoint callback, what are possible causes of this overhead?

Also, I am having difficulties in interpreting the performance difference between controllers, as in my understanding, the most costly negotiation should happen only once in the first epoch, and the negotiation results are cached in the rest of epochs. Therefore, if the response cache is large enough to store all tensors, in the remaining epochs, the controller only needs to update the response cache bit vector via MPI_Allreduce or Gloo_Allreduce, right? But I don't think MPI_Allreduce on the bit vector brings such overhead. Then what else operations are processed by the controller?
 
Specifically, I have fixed the steps count per epoch to 20, thus, the computation workload and communication frequency should remain unchanged in both senarios. However, the total amount of NCCL Allreduce operations is not a constant number, for either Gloo or MPI case. If we assume that the computation cost and NCCL operation cost are the same for both, why is there a hugh difference in controller operations?

The difference is more obvious when training a larger model.

ResNet50 test (batch size 32, 10 epochs per run):
|      | Total execution cost (s) |  NCCL Allreduce cost (ms) | NCCL Allreduce cnout |
| ------------- |:-------------:| :-------------: | :--------: |
| MPI (1st run)  | 137.73 | 62 | 194 |
| Gloo (1st run) |  98.89 |  50 | 64 |
| MPI (2nd run)  | 134.20 | 65 | 165 |
| Gloo (2nd run) |  92.69 | 42 | 69 |
| MPI (3rd run)  | 136.83 | 61 | 172 |
| Gloo (3rd run) | 98.75 |  46 | 65 |

The reproducer of test_mnist.py is here:
```
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from tensorflow.keras import backend as K
import math
import tensorflow as tf
import horovod.keras as hvd
import random
import numpy
import time
import os
import logging

os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
logging.getLogger("tensorflow").setLevel(logging.WARNING)

random.seed(42)
numpy.random.seed(42)
tf.random.set_seed(42)

# Horovod: initialize Horovod.
hvd.init()
rank = hvd.rank()

# Horovod: pin GPU to be used to process local rank (one GPU per process)
devices=tf.config.list_physical_devices('GPU')
ndevices=len(devices)
mylocaldevice = hvd.local_rank()
tf.config.set_visible_devices(devices[mylocaldevice], 'GPU')
logical_gpus = tf.config.list_logical_devices('GPU')
print(len(devices), "Physical GPUs,", len(logical_gpus), "Logical GPU")

batch_size = 128
num_classes = 10

# Horovod: adjust number of epochs based on number of GPUs.
#epochs = int(math.ceil(20 / hvd.size()))
epochs = 10

# Input image dimensions
img_rows, img_cols = 28, 28

# The data, shuffled and split between train and test sets
dpath = '/gpfs/alpine/scratch/jli111/csc312/mpi_dl/mnist.npz'
(x_train, y_train), (x_test, y_test) = mnist.load_data(dpath)

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
dsize = 1000
x_train = x_train[:dsize]
y_train = y_train[:dsize]
x_test = x_test[:dsize]
y_test = y_test[:dsize]

if rank == 0:
    print('x_train shape:', x_train.shape)
    print(x_train.shape[0], 'train samples')
    print(x_test.shape[0], 'test samples')

# Convert class vectors to binary class matrices
y_train = tf.keras.utils.to_categorical(y_train, num_classes)
y_test = tf.keras.utils.to_categorical(y_test, num_classes)

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

# Horovod: adjust learning rate based on number of GPUs.
opt = tf.keras.optimizers.Adadelta(0.1 * hvd.size())

# Horovod: add Horovod Distributed Optimizer.
opt = hvd.DistributedOptimizer(opt)

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=opt,
              metrics=['accuracy'])

callbacks = [
    # Horovod: broadcast initial variable states from rank 0 to all other processes.
    # This is necessary to ensure consistent initialization of all workers when
    # training is started with random weights or restored from a checkpoint.
    hvd.callbacks.BroadcastGlobalVariablesCallback(0),
]
## Timing echo epoch
class TimingCallback(tf.keras.callbacks.Callback):
    def __init__(self, logs={}):
        self.logs=[]
        self.epochs=[]
        self.curr_epoch = 0
    def on_train_begin(self, epoch, logs={}):
        self.starttime = time.time()
        t = time.time()
        if hvd.rank() == 0:
            print("\n### Training Start timestamp:", t)
    def on_epoch_begin(self, epoch, logs={}):
        self.es = time.time()
        self.curr_epoch += 1
    def on_epoch_end(self, epoch, logs={}):
        if hvd.rank() == 0:
            print("\n### Time from start:", time.time() - self.starttime, " epoch ", self.curr_epoch, " spent ", time.time() - self.es)

cb = TimingCallback()
callbacks.append(cb)

# Horovod: save checkpoints only on worker 0 to prevent other workers from corrupting them.
#if hvd.rank() == 0:
    #callbacks.append(keras.callbacks.ModelCheckpoint('./checkpoint-{epoch}.h5'))

tic = time.time()

model.fit(x_train, y_train,
          steps_per_epoch = 50,
          batch_size=batch_size,
          callbacks=callbacks,
          epochs=epochs,
          verbose=1 if hvd.rank() == 0 else 0,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)

toc = time.time()
if rank == 0:
    print('### Total training time:', toc-tic)
    print('Test loss:', score[0])
    print('Test accuracy:', score[1])
```

**Comment by maxhgerlach** [comment]
I think generally MPI is expected to be somewhat more efficient than Gloo. But I wonder if in your case, where NCCL handles the bulk, the difference be due to one-off costs like initialization. Have you tried some Python-level profiling to see where your script spends time? 

If you want to understand where time is spent and how well hardware is utilized, I've used Nvidia's NSight Systems profiler successfully in the past.

**Comment by romerojosh** [comment]
I think the difference here is due to variations in how Horovod is fusing tensors between the runs with the gloo and MPI controllers. This can be observed in the larger "NCCL count" in your tables. I think what is happening here is that MPI is running the control plane cycles faster than gloo. This causes the MPI runs to generate more individual (and small) allreduce communication calls than the gloo runs which seem to be reducing performance in your case. 

You can ad the `num_groups` argument to `DistributedOptimizer` to limit the number of allreduces generated by Horovod. I think this should recover performance in the MPI case.

**Comment by gseastream** [comment]
@jli111 did you ever figure out why MPI was slower than gloo? Or @romerojosh do you have any more tips? I am having the same problem on our cluster and I'm not sure why, using `horovod.tensorflow` on a large model.

I would have been happy to keep using gloo, but we use slurm as our HPC scheduler, and when gloo ssh's into other nodes it loses the virtual resource boundaries that slurm enforces so it uses GPUs and CPU cores and RAM that was not allocated to the job. But our mpi is slurm aware and doesn't have this problem, so I'd prefer to use mpi, but TF compilation is 99% slower and iteration time is 42% slower when the controller is mpi.

**Comment by gseastream** [comment]
Oh wow I figured it out 30 minutes later. For others finding this page, I increased the number of cpus available (per gpu) and now it's the same speed as gloo. I assume gloo was faster before because like I said, it ssh'ed into other nodes and was no longer limited to only 1 cpu per gpu by slurm (which is bad behavior because it's ignoring the scheduler).
